src/webextensions.c

```diff
@@ -0,0 +1,629 @@
+                            "app-id", xdp_app_info_get_id (call->app_info),
+                            "token", session_token,
+                            "connection", connection,
+                            NULL);
+
+  if (session)
+    g_debug ("webextensions session owned by '%s' created", session->sender);
+
+  return (WebExtensionsSession *)session;
+}
+
+static gboolean
+handle_create_session (XdpWebExtensions *object,
+                       GDBusMethodInvocation *invocation,
+                       GVariant *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(GError) error = NULL;
+  Session *session;
+
+  if (!xdp_app_info_is_web_browser (call->app_info))
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "cannot start WebExtensions native messaging servers");
+      return TRUE;
+    }
+
+  session = (Session *)web_extensions_session_new (arg_options, call, connection, &error);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+  if (!session_export (session, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      session_close (session, FALSE);
+      return TRUE;
+    }
+  session_register (session);
+
+  xdp_web_extensions_complete_create_session (object, invocation, session->id);
+
+  return TRUE;
+}
+
+static void
+on_server_exited (GPid pid,
+                  gint status,
+                  gpointer user_data)
+{
+  Session *session = user_data;
+  WebExtensionsSession *we_session = (WebExtensionsSession *)session;
+
+  SESSION_AUTOLOCK (session);
+  we_session->child_pid = -1;
+  we_session->child_watch_id = 0;
+  session_close (session, TRUE);
+}
+
+static gboolean
+array_contains (JsonArray *array,
+                const char *value)
+{
+  guint length, i;
+
+  if (array == NULL)
+    return FALSE;
+
+  length = json_array_get_length (array);
+  for (i = 0; i < length; i++)
+    {
+      const char *element = json_array_get_string_element (array, i);
+      if (g_strcmp0 (element, value) == 0)
+        return TRUE;
+    }
+  return FALSE;
+}
+
+static char *
+find_server (const char *server_name,
+             const char *extension_or_origin,
+             char **server_description,
+             GError **error)
+{
+  const char *hosts_path_str;
+  g_auto(GStrv) hosts_path;
+  g_autoptr(JsonParser) parser = NULL;
+  int i;
+
+  hosts_path_str = g_getenv ("XDG_DESKTOP_PORTAL_WEB_EXTENSIONS_PATH");
+  if (hosts_path_str == NULL)
+    {
+      hosts_path_str = "/usr/lib/mozilla/native-messaging-hosts:/etc/opt/chrome/native-messaging-hosts:/etc/chromium/native-messaging-hosts";
```

Comment:
Should this also include `/etc/opt/edge/native-messaging-hosts` for Microsoft Edge?

Reply:
I'm honestly not sure about this one. If we have any messaging hosts supporting Edge, would the json files they install for Chrome likely cover the IDs of the Edge extension too?